src/connection-pool-party.js

```diff
@@ -484,4 +516,86 @@ export default class ConnectionPoolParty extends EventEmitter {
         err => err,
       );
   }
+
+  _prioritizePools = () => {
+    if (!this.warmedUp) {
+      debug('_prioritizePools called before warmup completed. this should not happen.');
+      return;
+    }
+    if (this._healingPromise) {
+      debug('_prioritizePools called during heal, skipping this run');
+      return;
+    }
+    if (this.pools.length === 0) {
+      debug('_prioritizePools called when no pools exist, this should not happen');
+      return;
+    }
+    if (this.pools.length === 1) {
+      debug('_prioritizePools called when only one pool exists, skipping this run');
+      return;
+    }
+    if (this._prioritizePromise) {
+      debug('_prioritizePools called again while already in progress');
+    } else {
+      debug('_prioritizePools called');
+    }
+    const firstPoolPriority = this.pools[0].dsn.priority;
+    if (firstPoolPriority === undefined || firstPoolPriority === 0) {
+      debug('first pool has top priority, no need to prioritize');
+      return;
+    }
+    const higherPriorityPools = this.pools.filter(
+      pool => pool.dsn.priority < firstPoolPriority,
+    );
+    if (higherPriorityPools.length === 0) {
+      debug('unexpected priority config on DSNs, unable to prioritize');
+      return;
+    }
+    const unhealthyPriorityPools = higherPriorityPools.filter(
+      pool => !pool.connection.connecting && !pool.connection.connected,
+    );
+    this._prioritizePromise = this._prioritizePromise || Promise.resolve(unhealthyPriorityPools)
+      .then((unhealthyPools) => {
+        // If all the pools of a higher priority than index 0 are healthy, we can
+        // skip the heal. If there are any unhealthy pools,
+        // we need to heal them before sorting.
+        if (unhealthyPools.length === 0) {
+          this.pools.sort(poolPrioritySort);
+          debug('no unhealthy pools detected during prioritization');
+          return true;
+        }
+        debug(`healing ${unhealthyPools.length} pools during prioritization`);
+        return this._healPools(unhealthyPools);
+      })
+      .then((anyHealthyPools) => {
+        if (!anyHealthyPools) {
+          debug('none of the pools eligible for prioritization are healthy, unable to prioritize');
+          return;
+        }
+        // if any of the eligible pools are healthy, we prioritize. it would be preferable
+        // if we identified which pools didn't heal and ignore them in the
+        // prioritization, but that would require a little refactoring.
+        this.pools.sort(poolPrioritySort);
+        debug('prioritized pools');
+      })
+      .then(() => {
```

Comment:
Promise.finally for  ` this._prioritizePromise = null;` ?

Reply:
afaik Promise.finally just landed in v8 6.3, which is not part of Node.js 8

Reply:
yea, using native promises so no `finally` in older versions of node

Reply:
But the project is transpiled with Babel, probably includes the es6 promise polyfill

Reply:
good point, i'll see if it's supported

Reply:
didn't seem to like it, i'll just keep the then/catch duplication for now